contracts/airdrops.js

```diff
@@ -0,0 +1,244 @@
+        && api.assert(nativeTokenBalance
+          && api.BigNumber(nativeTokenBalance).gte(airdrop.quantity), 'you must have enough tokens to do the airdrop')) {
+        // validations completed
+        // lock airdrop tokens by transfering them to contract
+        const tokenTransfer = await api.executeSmartContract('tokens', 'transferToContract', {
+          to: CONTRACT_NAME, symbol, quantity: airdrop.quantity,
+        });
+
+        if (transferIsSuccesfull(tokenTransfer, 'transferToContract', api.sender, CONTRACT_NAME, symbol, airdrop.quantity)) {
+          // deduct fee from sender's utility token balance
+          const feeTransfer = await api.executeSmartContract('tokens', 'transfer', {
+            to: 'null', symbol: UTILITY_TOKEN_SYMBOL, quantity: airdrop.fee, isSignedWithActiveKey,
+          });
+
+          if (transferIsSuccesfull(feeTransfer, 'transfer', api.sender, 'null', UTILITY_TOKEN_SYMBOL, airdrop.fee)) {
+            const res = await api.db.insert('pendingAirdrops', {
+              airdropId: api.transactionId,
+              symbol,
+              type,
+              list: airdrop.list,
+              blockNumber: api.blockNumber,
+            });
+
+            api.emit('newAirdrop', { airdropId: res.airdropId });
+          } else {
+            // if fee transfer was failed, return native balance to api.sender
+            await api.transferTokens(api.sender, symbol, airdrop.quantity, 'user');
+          }
+        }
+      }
+    }
+  }
+};
+
+const processAirdrop = async (airdrop, maxTransactionsPerBlock) => {
+  const {
+    airdropId,
+    list,
+    symbol,
+    type,
+  } = airdrop;
+
+  let airdropIsPending = true;
+  let count = 0;
+
+  const completedDrops = [];
+
+  while (airdropIsPending) {
+    if (count < maxTransactionsPerBlock) {
+      if (list[count] !== undefined) {
+        const { to, quantity } = list[count];
+
+        if (type === 'transfer') {
+          // transfer tokens
+          await api.transferTokens(to, symbol, quantity, 'user');
+        } else if (type === 'stake') {
+          // stake tokens
+          await api.executeSmartContract('tokens', 'stakeFromContract', {
+            to, symbol, quantity,
+          });
+        }
+
+        completedDrops.push(list[count]);
+
+        count += 1;
+      } else {
+        // if list[count] is undefined, airdrop is finished
+        airdropIsPending = false;
+      }
+    } else {
+      airdropIsPending = false;
+    }
+  }
+
+  airdrop.list.splice(0, completedDrops.length);
+
+  if (airdrop.list.length > 0) {
+    // if limit has been reached & transactions are still remaining, update airdrop
+    await api.db.update('pendingAirdrops', airdrop);
+  } else {
+    // if no other transactions are remaining, delete airdrop
+    await api.db.remove('pendingAirdrops', airdrop);
+  }
+
+  api.emit('airdropDistribution', {
+    airdropId,
+    list: completedDrops,
+  });
```

Comment:
Perhaps you don't need the full completedDrops listed in this event. Just the airdrop ID and how many were processed would be enough. Also you will want to modify account history server later to trail these events and have these show up in the history.